tests/test_tagged_threads.cpp

```diff
@@ -0,0 +1,161 @@
+    lmq.job(sleeper);
+    lmq.job(sleeper);
+    // This one should stall:
+    std::atomic<bool> bad = false;
+    lmq.job([&] { bad = true; });
+
+    std::this_thread::sleep_for(50ms);
+
+    done = false;
+    lmq.job([&] { id = std::this_thread::get_id(); done = true; }, &t_abc);
+    wait_for([&] { return done.load(); });
+    {
+        auto lock = catch_lock();
+        REQUIRE( done.load() );
+        REQUIRE_FALSE( bad.load() );
+    }
+
+    done = false;
+    // We can queue up a bunch of jobs which should all happen in order, and all on the abc thread.
+    std::vector<int> v;
+    for (int i = 0; i < 100; i++) {
+        lmq.job([&] { if (std::this_thread::get_id() == id_abc) v.push_back(v.size()); }, &t_abc);
+    }
+    lmq.job([&] { done = true; }, &t_abc);
+    wait_for([&] { return done.load(); });
+    {
+        auto lock = catch_lock();
+        REQUIRE( done.load() );
+        REQUIRE_FALSE( bad.load() );
+        REQUIRE( v.size() == 100 );
+        for (int i = 0; i < 100; i++)
+            REQUIRE( v[i] == i );
+    }
+    sleep = false;
+    wait_for([&] { return bad.load(); });
+    {
+        auto lock = catch_lock();
+        REQUIRE( bad.load() );
+    }
+}
+
+TEST_CASE("batch job completion on tagged threads", "[tagged][batch-completion]") {
+    lokimq::LokiMQ lmq{get_logger(""), LogLevel::trace};
+
+    lmq.set_general_threads(4);
+    lmq.set_batch_threads(4);
+    std::thread::id id_abc;
+    auto t_abc = lmq.add_tagged_thread("abc", [&] { id_abc = std::this_thread::get_id(); });
+    lmq.start();
+
+    lokimq::Batch<int> batch;
+    for (int i = 1; i < 10; i++)
+        batch.add_job([i, &id_abc]() { if (std::this_thread::get_id() == id_abc) return 0; return i; });
+
+    std::atomic<int> result_sum = -1;
+    batch.completion([&](auto result) {
+        int sum = 0;
+        for (auto& r : result)
+            sum += r.get();
+        result_sum = std::this_thread::get_id() == id_abc ? sum : -sum;
+    }, &t_abc);
+    lmq.batch(std::move(batch));
+    wait_for([&] { return result_sum.load() != -1; });
+    {
+        auto lock = catch_lock();
+        REQUIRE( result_sum == 45 );
+    }
+}
+
+
+TEST_CASE("timer job completion on tagged threads", "[tagged][timer]") {
+    lokimq::LokiMQ lmq{get_logger(""), LogLevel::trace};
+
+    lmq.set_general_threads(4);
+    lmq.set_batch_threads(4);
+
+    std::thread::id id_abc;
+    auto t_abc = lmq.add_tagged_thread("abc", [&] { id_abc = std::this_thread::get_id(); });
+    lmq.start();
+
+    std::atomic<int> ticks = 0;
+    std::atomic<int> abc_ticks = 0;
+    lmq.add_timer([&] { ticks++; }, 10ms);
+    lmq.add_timer([&] { if (std::this_thread::get_id() == id_abc) abc_ticks++; }, 10ms, true, &t_abc);
+
+    wait_for([&] { return ticks.load() > 2 && abc_ticks > 2; });
+    {
+        auto lock = catch_lock();
+        REQUIRE( ticks.load() > 2 );
+        REQUIRE( abc_ticks.load() > 2 );
+    }
+}
+
+
```

Comment:
maybe this is covered, but some tests around missing, duplicate, etc. tagged threads would be good

Reply:
I don't know what missing or duplicate means; nothing prevents you from creating two tagged threads with the same name, and you don't have to keep the reference to the thread -- you could, for example, set up a thread to handle timers, register the timers, then drop the TaggedThread.

Reply:
I redid this a bit and renamed TaggedThread to TaggedThreadID to make it a bit clearer what you actually have.  Also dropped the string so it's just a single int wrapper and thus cheaply passable, so now it gets passed by value (with std::optionals) instead of by nullable pointer.